### PR TITLE
[SandboxVec][Scheduler] Implement rescheduling

### PR DESCRIPTION
**Pull request by vporpo**
This patch adds support for re-scheduling already scheduled instructions. For now this will clear and rebuild the DAG, and will reschedule the code using the new DAG.

### CODE REVIEW

**Comment by vporpo** [comment]
Fixed use-after-free caused by DAG.clear().

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Scheduler.h [hunk 1]
```diff
@@ -94,19 +104,30 @@ class Scheduler {
...
   bool tryScheduleUntil(ArrayRef<Instruction *> Instrs);
   /// Schedules all nodes in \p Bndl, marks them as scheduled, updates the
   /// UnscheduledSuccs counter of all dependency predecessors, and adds any of
   /// them that become ready to the ready list.
   void scheduleAndUpdateReadyList(SchedBundle &Bndl);
-
+  /// The scheduling state of the instructions in the bundle.
+  enum class BndlSchedState {
+    NoneScheduled,
+    PartiallyOrDifferentlyScheduled,
```
Comment (slackito) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Scheduler.h:124]: "Partially" seems obvious to me (only some instructions in the bundle have been scheduled so far) but not so for "Differently". Could you add a comment?

**Review by slackito — COMMENTED** [review]
I'll finish reviewing this later. Just a quick comment for now.

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Scheduler.h (hunk 1 above)
Comment (vporpo) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Scheduler.h:124]: "Differently" is when the instructions belong to different bundles, which suggests that they were previously scheduled in a way that differs from the way we attempt to schedule them now (that is all instructions in a single bundle). I will add some comments.

**Review by slackito — APPROVED** [review]
Other than the previous comment, looks good to me.

**Pull request merged**